### DISCUSSION TITLE
How to Style? Best practices and recommendations?

### DISCUSSION THREAD

**Discussion by dggsax** [microsoft/fluentui#20170] [General] [answered]
There seems to be quite a few ways to handle styling. We can create the styles in-line with the component props. There's the new `makeStyles` hook (which is very cool, by the way), and then there's the `mergeStyles` method as well. From the FluentUI team, are there any resources that break down the pros and cons of either approach and, better yet, is there a recommendation for the best way to go? Thanks in advance, look forward to learning more about this space 😀

**Comment by dkozar** [comment, marked as answer]
It's important to understand the difference between `style`, `styles `and `className `props.

CSS in JS (`mergeStyles`, `mergeStyleSets`) is used with `className` prop, whereas `style` and `styles` are _inline styles_, applied directly to [DOM element's style property](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/style). 
 
CSS in JS **is a true CSS**, meaning it injects CSS style declarations to document's `<head>`.

With CSS in JS you could use most of the CSS features like cascading etc.

## Quick demo

To understand what CSS in JS does:
 
```js
/* copy-paste to your browser console and hit Enter */ 
var style = document.createElement('style'); 
style.innerHTML = "div {border: 5px solid black; background-color: blue;}"; 
document.head.appendChild(style); 
```

If you go (for example) to https://www.microsoft.com/en-us/, and paste this snippet to browser console, background of all the divs in the page will become blue:

![image](https://user-images.githubusercontent.com/1078403/141281237-5aeebd11-558a-4e76-8b60-f72817f81f0d.png)

## Actual implementation

Here's how [mergeStyleSets](https://docs.microsoft.com/en-us/javascript/api/merge-styles?view=office-ui-fabric-react-latest#mergeStyleSets_styleSet_) is implemented:

1. It **receives a map** (of _CSS style rule objects_ as values) as an argument:

```js
{
    foo: {
        color: 'blue'
    },
    bar: {
        color: 'red'
    }
}
```

2. It **generates a different map** of randomized unique **strings** as values, and returns it to be used by your component:

```js
{
    foo: '3fh9483hrf4ioewjzour', // string
    bar: 'dfsjk43iow3rjf9ejwfjf0' // string
}
```

Your component:

```js
<Component className={styles.foo} />
```

So effectivelly:

```js
<Component className="3fh9483hrf4ioewjzour" />
```

This adds the "3fh9483hrf4ioewjzour" class name (string) to component's DOM element:

```html
<div class="3fh9483hrf4ioewjzour" />
```

3. It **creates a CSS style sheet** and pushes it to the page `<head>`:

```js
var style = document.createElement('style'); 

style.innerHTML = `
.3fh9483hrf4ioewjzour {
    color: blue;
}
.dfsjk43iow3rjf9ejwfjf0{
    color: red;
}`; 

document.head.appendChild(style); 

```

It creates it only once - first time the component is mounted.

**Reply by dggsax** [reply]
Thank you very much for the breakdown! Since writing this I have experimented more with the `mergeStyleSets` approach for styling and it seems pretty great. I ended up having to utilize a pattern for Function Components to make sure the styles were always updated in case the selected theme changed for the overall component:

```tsx

// Simple function definition so we can pass in the theme and create the styles
const getStyles = (theme: ITheme) => 
    mergeStyleSets({
        text: {
            color: theme.palette.themePrimary
        }
    });

const FunctionComponent = () => {
    // Use the hook instead of getTheme so when we change the theme provided to ThemeProvider we receive the updates
    // At least I think that's how this is working...
    const theme = useTheme();
    const styles = getStyles(theme);

    return <Text variant="xLarge" className={styles.text}>Hello Github!</Text>;
}
```

This seems to be working well for us thus far, I'm just unsure if there are any tradeoffs for this approach.

One thing about `mergeStyleSets` that I don't like so far compared to using the `styles` prop directly is that you don't get the autocomplete CSS properties. For instance, if I have the following:
```tsx
const example = mergeStyleSets({
    thing: {
        ba
    }
```
I would expect to see VSCode have the intellisense options with things like "background" or "backgroundColor" show up, but that isn't the case.

**Reply by dkozar** [reply]
Well, I have the intelissense (of style props). I'd assume it's just a metter of configuration.

**Reply by dkozar** [reply]
Furthermore, if you want to ensure the best performance with mergeStyleSets, you could wrap it into Fluent's memoizeFunction:

https://github.com/microsoft/fluentui/blob/master/packages/merge-styles/README.md#optimizing-for-performance

**Reply by dkozar** [reply]
@dggsax to get the Intellisense, you should type each selector as `IStyle`:

```js
import { IProcessedStyleSet, IStyle, memoizeFunction, mergeStyleSets } from "@fluentui/react";

interface YourStyles {
  root: IStyle;
  child: IStyle;
}

export const getClassNames = memoizeFunction(
  (): IProcessedStyleSet<YourStyles> =>
    mergeStyleSets({
      root: {
        // intellisense here
      },
      child: {
        // intellisense here
      },
    })
);
```

![image](https://user-images.githubusercontent.com/1078403/149130423-4ea1dd6f-b6f6-4a45-b34a-a53e43245af5.png)

![image](https://user-images.githubusercontent.com/1078403/149130447-c7d2acbd-50ae-4e00-b781-433692cc83b3.png)
